baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/BuilderPrivateConstructor.java

```diff
@@ -0,0 +1,108 @@
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.Tree.Kind;
+import com.sun.tools.javac.code.Symbol.ClassSymbol;
+import com.sun.tools.javac.code.Symbol.MethodSymbol;
+import com.sun.tools.javac.tree.JCTree;
+import java.util.List;
+import java.util.Set;
+import javax.lang.model.element.Modifier;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "BuilderPrivateConstructor",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        providesFix = BugPattern.ProvidesFix.REQUIRES_HUMAN_ATTENTION,
+        severity = BugPattern.SeverityLevel.WARNING,
+        summary = "A Builder with a static factory method on the encapsulating class must have a private constructor. "
+                + "Minimizing unnecessary public API prevents future API breaks from impacting consumers. ")
+public final class BuilderPrivateConstructor extends BugChecker implements BugChecker.ClassTreeMatcher {
+
+    @Override
+    public Description matchClass(ClassTree tree, VisitorState state) {
+        ClassSymbol enclosingClass = ASTHelpers.enclosingClass(ASTHelpers.getSymbol(tree));
+        if (enclosingClass == null) {
+            return Description.NO_MATCH;
+        }
+        if (!isValidBuilderClass(tree)) {
+            return Description.NO_MATCH;
+        }
+        List<MethodTree> constructors = ASTHelpers.getConstructors(tree);
+        if (constructors.size() != 1 || !ASTHelpers.isGeneratedConstructor(Iterables.getOnlyElement(constructors))) {
+            return Description.NO_MATCH;
+        }
+
+        if (!hasStaticBuilderFactory(enclosingClass, state)) {
+            // No factory method, the public constructor is used
+            return Description.NO_MATCH;
+        }
+
+        // If no fields exist, the constructor is placed after the curly brace
+        int constructorPosition = ((JCTree) tree).getStartPosition()
+                + state.getSourceForNode(tree).indexOf('{')
+                + 1;
+
+        for (Tree member : tree.getMembers()) {
+            if (member.getKind() == Kind.VARIABLE) {
+                int endPosition = state.getEndPosition(member);
+                if (endPosition > constructorPosition) {
+                    constructorPosition = endPosition;
+                }
+            }
+        }
+        return buildDescription(tree)
+                .addFix(SuggestedFix.builder()
+                        .replace(
+                                constructorPosition,
+                                constructorPosition,
+                                String.format("\nprivate %s() {}", tree.getSimpleName()))
+                        .build())
+                .build();
+    }
+
+    private static boolean isValidBuilderClass(ClassTree tree) {
+        return tree.getSimpleName().contentEquals("Builder")
+                && tree.getImplementsClause().isEmpty()
```

Comment:
requiring it to not implement any methods is probably going to mean we miss any staged builders right? (these would have one class implementing `FooStage`, `BarStage`, `BazStage` which each declare one method?)

Reply:
That's correct, however staged builders generally use several interfaces with an internal hidden implementation.